### ISSUE TITLE
Question about the Performance on NOCS-REAL275 (Tab 1 & Tab 8)

### ISSUE THREAD

**Issue by taeyeop-lee** [HalfSummer11/CAPTRA#9]
Thanks for sharing good work.

I have some questions related to the performance on NOCS-REAL275 (Tab 1 & Tab 8)

Q1. Can I ask why the reported CASS performance of 5°, 5cm is different compared to the original CASS paper (23.5 vs 29.44)?? Can you explain how to get these results including mIoU, R_err, and T_err metrics?? Also, the reported 6-PACK performance of  5°, 5cm is worse than the original paper (33.3 vs 28.92).

Q2. Can you explain the Oracle ICP and what is different the ICP from 6-PACK?? The paper 6-PACK also reported the ICP and it performs close to NOCS at 5°, 5cm metric. Can you explain why your Oracle ICP is worse than NOCS??

Q3. As I understand your method uses the same random noise (scale = 0.02, rot = 5°, trans=3cm) with ground-truth initial pose during training and inference. Have you tried different parameter settings (ex, rot= 5°, trans=5cm)?? Also, Can you explain where the code is implemented related to rotation noise?? I only found the translation and scale noisy part of the below code.
https://github.com/HalfSummer11/CAPTRA/blob/d98158222f5f6bc44687d75309caf03300e185e1/datasets/nocs_data/nocs_data_process.py#L31

**Comment by HalfSummer11** [comment]
Hi, thanks for your interest in our work! Regarding your questions,

A1: CASS and NOCS are detection-based methods, where multiple predictions may exist for the same object. Therefore they used Average Precision when they compute 5º, 5cm of their predictions. As a tracking-based method, we follow 6-PACK and directly compute 5º, 5cm (as well as mIoU, R_err, T_err) by taking the average over all predictions - since we have exactly one prediction per object. For a fair comparison, we apply the same metric to CASS's raw prediction results. That's why we have 29.44 (mean) instead of 23.5 (mean Average Precision). 

For 6-PACK in the "perturb" setting, we use the same initial error distribution as ours, i.e. including both rotation error and translation error. As we already noted in the paper, 6-PACK originally only initializes with translation errors. This may be the reason of their degraded performance.

A2: We used a self-implemented ICP where different choices of parameters/implementation details may impact the result. 6-PACK probably used another implementation. 

A3: We tried different parameter settings in some small-scale experiments. In general, a larger noise scale makes the trained model more robust at the cost of less precise predictions. 
Noise is added to the pose [here](https://github.com/HalfSummer11/CAPTRA/blob/d98158222f5f6bc44687d75309caf03300e185e1/network/models/model.py#L394) during training/inference. The code you referenced is only used to perturb the ball we cropped from the scene point cloud. As it's a ball, rotation doesn't matter.

**Comment by taeyeop-lee** [comment]
Thanks for the quick reply. 
I have some additional follow questions related to your answers.

Q1. I understand CASS and NOCS case. I agree that your "perturb", which considers rotation and translation, makes degrade the performance of the original 6-PACK. However, It is still hard to understand the 6-PACK "GT" case. The performance (28.92) is lower than only considering translation noise (33.3). Can you share your opinion related to this result??

Q2. Can you share your opinion that why ICP shows poor performance even using the 3D CAD models?? 

Q3. Does it mean that if you use a larger noise scale during training, expect more accurate performance than small scale noise when the same noise is given during inference?? I also wonder if the method you proposed fails to track when given how much perturb. This information would be really helpful to the community.

**Comment by HalfSummer11** [comment]
A1: Based on my observation, for tracking methods like 6-PACK, errors may accumulate as the tracking goes on. Also once they lose track, the error can become very large - as you can see in the supplementary videos. This is the main source of error. How large this error is may not be in proportion to the amount of error in the initialization (like after 10 frames we probably have the same amount of accumulated error no matter how we initialized. Exception: initial error is already too large for the network to handle).

A2: I'm not an expert with ICP-based tracking - I never expect it to work well, either. My basic understanding is ICP may work well when the two shapes to be matched are clean, complete, and have very close poses (+ a good implementation with the right parameters that make the optimization converge). But here the shapes are partial, noisy, and the initial poses are not guaranteed to be close enough. 

A3: I'm not sure what you mean by "the same noise". If "the same noise" is the larger noise scale - then maybe. Because a network trained with small-scale noise may not be able to handle such large initial noise. Otherwise, I suspect not. Because training with larger noise contributes to *robustness* and probably *hurts accuracy*.  
Please refer to Table 11 to see our method's robustness to pose perturbation. We didn't run a stress test, though. You may play with our pretrained model to find out more.

**Comment by taeyeop-lee** [comment]
Sorry for the late reply. 
I have some follow-up questions related to A1.

Q1. As I understand, there are two main reasons that different results (6-PACK and CAPTRA) for each inference. First, the random perturbation would make different initialization, and it affects different results. Second, the algorithm is still not repeatable even if the initialization is the same. Can I understand that 6-PACK and your method do not show repeatable results, even if your initialization is the same condition? 
- https://github.com/j96w/6-PACK/issues/19#issuecomment-674659170


Q2. Can you explain what is the main difference between the [video](https://www.youtube.com/watch?v=EkcCEj7gZGg) and your scripts?? I checked [your video](https://www.youtube.com/watch?v=EkcCEj7gZGg) comparing the result with 6-PACK. Also, I followed your provided scripts to visualize the results. Specifically, the predictions of the camera usually fail to track, when I followed your scripts. I share some results from your scripts to understand. I was also surprised that the first camera prediction track the wrong class (mug) instead of the camera. Have you also observed these results??


![image](https://user-images.githubusercontent.com/50736858/150080095-abb891f3-29ec-4152-aece-c1ef8694204b.png)
![image](https://user-images.githubusercontent.com/50736858/150080146-2d25b0ba-362d-4903-acac-b1583c098d09.png)

**Comment by HalfSummer11** [comment]
Hi, regarding Q1, yes, our program is not purely deterministic, meaning there're random factors, like the first point sampled during FPS, that affect the predictions (this is something internal in PointNet++, also something that will directly change the input point cloud to our model). The fact that this is a tracking model where each frame relies on the previous frame's prediction magnifies this random factor - since the error may accumulate. 
Regarding Q2, our model does have difficulty tracking cameras on NOCS-REAL275. This is reflected in the numerical results we reported in the paper (see supp. for a per-category breakdown). We also extensively discussed this issue in the paper & supp. In our case, once the model loses track, the cropped input point cloud to the next frame may not include the correct object, further confusing the model. This is probably what's happening with the mug.

**Comment by taeyeop-lee** [comment]
Thanks for Q1-A1 and sorry for the unclear question of Q2!
For a supplementary explanation on Q2, My question is that your scripts results and your [supplementary video](https://www.youtube.com/watch?v=EkcCEj7gZGg) results are quite different for all classes, specifically, the camera class. In the supplementary video, your method predicts quite well to camera class, also other class also looks better than one. Is there any experimental difference between your script's results and supplementary video?? 
(Ex, Initial pose [scale, rotation, translation], was ground truth for supplementary video and scripts are initialized with noise)

![Compare](https://user-images.githubusercontent.com/50736858/150276104-7f9ff434-a7ef-400c-a304-093b936667e5.png)
